confmap/converter/expandconverter/expand.go

```diff
@@ -4,52 +4,68 @@
 package expandconverter // import "go.opentelemetry.io/collector/confmap/converter/expandconverter"
 
 import (
+    "fmt"
 	"context"
 	"os"
 
 	"go.opentelemetry.io/collector/confmap"
+	"go.uber.org/zap"
 )
 
-type converter struct{}
+type converter struct{
+    logger *zap.Logger
+
+    // Record of which env vars we have logged a warning for
+    loggedDeprecations map[string]bool
+}
 
 // New returns a confmap.Converter, that expands all environment variables for a given confmap.Conf.
 //
 // Notice: This API is experimental.
 func New(_ confmap.ConverterSettings) confmap.Converter {
-	return converter{}
+	return converter{
+        loggedDeprecations: make(map[string]bool),
+        logger: zap.NewNop(), // TODO: pass logger in ConverterSettings
+    }
 }
 
-func (converter) Convert(_ context.Context, conf *confmap.Conf) error {
+func (c converter) Convert(_ context.Context, conf *confmap.Conf) error {
 	out := make(map[string]any)
 	for _, k := range conf.AllKeys() {
-		out[k] = expandStringValues(conf.Get(k))
+		out[k] = c.expandStringValues(conf.Get(k))
 	}
 	return conf.Merge(confmap.NewFromStringMap(out))
 }
 
-func expandStringValues(value any) any {
+func (c converter) expandStringValues(value any) any {
 	switch v := value.(type) {
 	case string:
-		return expandEnv(v)
+		return c.expandEnv(v)
 	case []any:
 		nslice := make([]any, 0, len(v))
 		for _, vint := range v {
-			nslice = append(nslice, expandStringValues(vint))
+			nslice = append(nslice, c.expandStringValues(vint))
 		}
 		return nslice
 	case map[string]any:
 		nmap := map[string]any{}
 		for mk, mv := range v {
-			nmap[mk] = expandStringValues(mv)
+			nmap[mk] = c.expandStringValues(mv)
 		}
 		return nmap
 	default:
 		return v
 	}
 }
 
-func expandEnv(s string) string {
+func (c converter) expandEnv(s string) string {
 	return os.Expand(s, func(str string) string {
+        if (!c.loggedDeprecations[str]) {
+            msg := fmt.Sprintf("Variable substitution using $VAR will be deprecated in favor of ${env:VAR}, please update $%v", str)
+            c.logger.Warn(msg, zap.String("variable", str))
+            c.loggedDeprecations[str] = true
```

Comment:
As you can see on https://pkg.go.dev/os#Expand, 
> Expand replaces ${var} or $var in the string based on the mapping function.

We only want to log on the `$var` case, but not on the `${var}` case. An easy way to do this is to use a regexp and filter out the instances of `${..}`. You could also have a look at the source code for `os.Expand` and modify accordingly

Reply:
I understand, have not looked too deep at the code path and why it is that way, but when I used ${env:VAR} that code was not reached at all. Maybe you could double check for me, I wanted to write the tests for it but had the challenges described in the PR.

Reply:
I see, I totally missed that ${var} was a thing, thought it was only ${env:var}. I have used regex to find $var specifically instead of ${var}. because a given line could have a combination of $VAR and ${VAR}, so we need to look for $VAR inside it. Let me know if this logic is off